src/realm/sync/noinst/protocol_codec.hpp

```diff
@@ -491,6 +494,63 @@ class ClientProtocol {
         return action::ApplicationBug;
     }
 
+    template <typename Connection>
+    void parse_log_message(Connection& connection, HeaderLineParser& msg)
+    {
+        auto report_error = [&](const auto fmt, auto&&... args) {
+            auto msg = util::format(fmt, std::forward<decltype(args)>(args)...);
+            connection.handle_protocol_error(Status{ErrorCodes::SyncProtocolInvariantFailed, std::move(msg)});
+        };
+
+        auto session_ident = msg.read_next<session_ident_type>();
+        auto message_length = msg.read_next<size_t>('\n');
+        auto message_body_str = msg.read_sized_data<std::string_view>(message_length);
+        nlohmann::json message_body;
+        try {
+            message_body = nlohmann::json::parse(message_body_str);
+        }
+        catch (const nlohmann::json::exception& e) {
+            return report_error("Malformed json in log_message message: \"%1\": %2", message_body_str, e.what());
+        }
+        static const std::unordered_map<std::string_view, util::Logger::Level> name_to_level = {
+            {"fatal", util::Logger::Level::fatal},   {"error", util::Logger::Level::error},
+            {"warn", util::Logger::Level::warn},     {"info", util::Logger::Level::info},
+            {"detail", util::Logger::Level::detail}, {"debug", util::Logger::Level::debug},
+            {"trace", util::Logger::Level::trace},
+        };
+
+        // See if the log_message contains the appservices_request_id
+        if (auto it = message_body.find("co_id"); it != message_body.end() && it->is_string()) {
+            connection.receive_appservices_request_id(it->get<std::string_view>());
+        }
+
+        std::string_view log_level;
+        bool has_level = false;
```

Comment:
(why) is this optional?

Reply:
Kiro had suggested it could default to a value (debug) if it wasn't provided.